src/hotspot/share/opto/loopopts.cpp

```diff
@@ -3774,6 +3777,10 @@ bool PhaseIdealLoop::partial_peel( IdealLoopTree *loop, Node_List &old_new ) {
                             n->_idx, get_ctrl(n)->_idx);
             }
 #endif
+          } else if (n->Opcode() == Op_ScopedValueGetHitsInCache) {
+            // ScopedValueGetLoadFromCache and companion ScopedValueGetHitsInCacheNode must stay together
+            move_scoped_value_nodes_to_not_peel(peel, not_peel, peel_list, sink_list, i);
+            incr = false;
```

Comment:
Could we otherwise exhaust the node limit, by peeling a loop that is too large?

Reply:
No node is cloned here so there's no need to adjust the `estimate`. What happens is that a `ScopedValueGetHitsInCacheNode` is in the peeled region of the loop but not its `ScopedValueGetLoadFromCache` because peeling happens right above the `If` for the `ScopedValueGetHitsInCacheNode` . It's correct to simply move the `ScopedValueGetHitsInCacheNode`  out of the peeled region into the non peeled region because it's only used there.

There was no test case for this. I added one.